### ISSUE TITLE
Q: Are addresses signed or unsigned? ... for the purposes of address range CMOs...  (A: unsigned)

### ISSUE THREAD

**Issue by AndyGlew** [AndyGlew/Ri5-stuff#6]

**Comment by AndyGlew** [comment]
Most people assume that addresses are unsigned. However, canonicalization of addresses amounts to sign extending from the topmost implemented bit for the rest of the width. 

Address range instructions and/or the software loops that they are embedded within often require a determination as to whether addresses are signed or unsigned.

Consider the reference software loop or an address range instruction as proposed for RISC-V, assuming the [lo,hi) range representation:

~~~
     X1 := lo address
     X2 := hi address
1:  ARI X1,X1,X2
     BLTU X1,X2,1b
~~~
The loop branch must be BLT or BLTU depending on whether addresses are treated as signed or unsigned. Similarly, the address range instruction, which I am calling by nickname ARI, performs the same computation.

**Comment by AndyGlew** [comment]
It is possible to define address range instructions that do not require defining whether addresses are signed or unsigned.

For example, the instructions could be defined have a bite count, nbytes, and to decrement that byte count until the value 0 is reached. This works with either of the candidate nbytes range instruction varieties, XLEBN/2 and [hi-nbytes,hi)

Similarly, even [lo,hi) formats could be defined to as if the scan address starts at low, and is incremented until eventually it matches hi.  If the address space were fully populated, this would eventually wrap both signed and unsigned. It would correctly cross the -1,0 boundary as well as the 00..0,10..0 boundary.

The software loop could be

~~~
     X1 := lo address
     X2 := hi address
1:  ARI X1,X1,X2
     BNE X1,X2,1b
~~~

The issue of signed versus unsigned only comes up because

a) Mainly, because cache line oriented implementations of range instructions will return the address of the next cache line. They are permitted to ignore the RS2 input. Therefore, they may have passed the RS2 input, the high value. In particular, they may have crossed either the -1,0 boundary or the 0..0,10..0 boundary.  If the software loops are not to be complicated, a less than comparison is required, whether signed or unsigned.

b) certain varieties of the address range instructions are enabled by cramming extra information into the operands. E.g. see Issue #2, relating to allowing an address range instruction to be a no-op. This is no longer an active consideration because our POR has resolved issue #2 by saying that address range instructions cannot be NOPs - largely because the [lo,hi) form is ill-suited. (Both of the nbytes forms supports no-op-ing address range instructions.)

**Comment by AndyGlew** [comment]
RESOLUTION: for these purposes addresses are defined to be unsigned.

The reference software loop around an address range instruction "ARI" is
~~~
     X1 := lo address
     X2 := hi address
1:  ARI X1,X1,X2
     BLTU X1,X2,1b
~~~

Such instructions are not allowed to cross the -1,0 boundary, since that would imply that the high address is less than the low address when these addresses are treated as unsigned quantities.

Q: should the address range instructions trap when crossing the -1, 0 boundary? TBD: file another issue.

Cache line oriented address range instructions are not required to return a final RD to the upper address RS2. They may return the address of the next cache line past that upper address.
